### PR TITLE
add mkldnn fluid design doc

### LINKED ISSUE #6815: Discussion on integrating MKLDNN to fluid
**1. Background and consensus**
- An operator can have several kernels. And fluid provides a policy to switch the right kernel for an operator. Please refer to [switching kernel doc](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/switch_kernel.md)
- Since data layout will bring branches in computation, we need to record layout in Tensor for convenience. Otherwise, we have to expose Layout attribute to users. Please refer to #6765 

**2. concerns on Layout as the key of OpKernelType**

The essence about taking layout as the key OpKernelType is to write following codes in which part. 
- We can get proper Layout in `GetExpectedKernelType` method and convert data in `Trans` method, and then chose right kernel registered in advance.
- We can also put these codes inside the `Compute` method of MKLDNN kernel. The layout will be chosen and transformed inside kernel.

@tensor-tang Could you give some advice on the total effort of these two choice considering integrating MKLDNN to fluid.

**3. Some other related question**

Some operators like dropout and batch norm will have different computation logic in train/test process. Fluid handles these logic inside the `Compute` method of operator kernel. Fluid does not take `is_test` as a key of OpKernelType currently.

### LINKED ISSUE #6846: Add design doc for fluid on MKLDNN

### PR DESCRIPTION
**Pull request by tensor-tang**
fix #6846 

And this should also resolve #6815 and #6822 

click [here](https://github.com/tensor-tang/Paddle/blob/design_doc/doc/design/mkl/mkldnn_fluid.md) to view.

### CODE REVIEW

File: doc/design/mkl/mkldnn_fluid.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
+# Design Doc: Add MKLDNN Kernel in Fluid Operator
+
+
+## Principles
+
+First of all, we should follow some basical principles like:
+1.  [How to write a new operator](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/howto/dev/new_op_en.md). We are trying to add a new kind of kernel into operators, so basically we should follow this doc.
+2.  [Supporting new Device/Library](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/support_new_device.md). Since MKLDNN is a new library to fluid, we should add `MKLDNNDeviceContext` and maybe `mkldnn_helper.h`, just like [cudnn_helper.h](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/platform/cudnn_helper.h).
+3.  [Switch Kernel](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/switch_kernel.md). Another important point is that we should ensure the data synchronization among different divices, which is the topic #6549. So basically we should override `GetActualKernelType` and `trans` functions to support switching kernels.
```
Comment (QiJune) [inline doc/design/mkl/mkldnn_fluid.md]: 1. Actually, we should ensure data synchronization between different kernel types.

2. We should override `GetExpectedKernelType`.

File: doc/design/mkl/mkldnn_fluid.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+4.  [The Keys of Operator Kernel Type](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/operator_kernel_type.md). Kernel Type is a pivotal conception which can record the `Place`, `Library`, `DataType` and `Layout`.
+
+## Sulution
+
+In general, there are three part we should follow to run a MKL-DNN primitive.
```
Comment (QiJune) [inline doc/design/mkl/mkldnn_fluid.md]: Could you please give some background about MKLDNN primitive? What's a primitive? Is it a job in MKLDNN?

File: doc/design/mkl/mkldnn_fluid.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+-  create a primitive descriptor that describe this operator
+-  create a memory that handle all memory buffers needed
+-  stream the primitive create by first two
```
Comment (QiJune) [inline doc/design/mkl/mkldnn_fluid.md]: Launch a stream run the primitive created by first two?

File: doc/design/mkl/mkldnn_fluid.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+2.  [Supporting new Device/Library](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/support_new_device.md). Since MKLDNN is a new library to fluid, we should add `MKLDNNDeviceContext` and maybe `mkldnn_helper.h`, just like [cudnn_helper.h](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/platform/cudnn_helper.h).
+3.  [Switch Kernel](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/switch_kernel.md). Another important point is that we should ensure the data synchronization among different divices, which is the topic #6549. So basically we should override `GetActualKernelType` and `trans` functions to support switching kernels.
+4.  [The Keys of Operator Kernel Type](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/operator_kernel_type.md). Kernel Type is a pivotal conception which can record the `Place`, `Library`, `DataType` and `Layout`.
+
+## Sulution
+
+In general, there are three part we should follow to run a MKL-DNN primitive.
+-  create a primitive descriptor that describe this operator
+-  create a memory that handle all memory buffers needed
```
Comment (QiJune) [inline doc/design/mkl/mkldnn_fluid.md]: memory handle memory? Could you describe it more accurate?

File: doc/design/mkl/mkldnn_fluid.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+-  stream the primitive create by first two
+
+We do not want to see the first two would be re-initialized every iteration again and again. \
```
Comment (QiJune) [inline doc/design/mkl/mkldnn_fluid.md]: It's better to avoid reinitialization of primitive descriptor and memory handle in the first two stages in every iteration.

File: doc/design/mkl/mkldnn_fluid.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+So we plan to create a map to record all the `primitive` and `memory`, which should not take too much memories as discussed [here](https://github.com/PaddlePaddle/Paddle/issues/6822).
+
+Assuming that three condition would be confirmed:
```
Comment (QiJune) [inline doc/design/mkl/mkldnn_fluid.md]: It's assumed that following three conditions should be satisfied.

File: doc/design/mkl/mkldnn_fluid.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+1. there is a unique key for each operator instance. May be the actual name of `Output Tensor`.
+2. the `Input Tensor` inside `Compute` function is the one after converted.
+3. we can get the phase(eg. `is_test`) inside `Compute` function, otherwise we need to expose this attribue to user.
```
Comment (QiJune) [inline doc/design/mkl/mkldnn_fluid.md]: Please add some background, why we need `is_test` flags in MKLDNN

File: doc/design/mkl/mkldnn_fluid.md [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+### Compute
+The algorithm of `Compute` would be described as follow, let's take conv like an example.
+
+```c++
+
+  PADDLE_ENFORCE(platform::is_cpu_place(ctx.GetPlace()), "It must use CPUPlace.");
```
Comment (QiJune) [inline doc/design/mkl/mkldnn_fluid.md]: ctx here is actually ExecutionContext

File: doc/design/mkl/mkldnn_fluid.md [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+  PADDLE_ENFORCE(platform::is_mkldnn_library(ctx.GetLibrary()), "It must use MKLDNN Library.");
+
+  auto& ctx = executionContext.template device_context<platform::MKLDNNDeviceContext>();
```
Comment (QiJune) [inline doc/design/mkl/mkldnn_fluid.md]: ctx here is actually DeviceContext

File: doc/design/mkl/mkldnn_fluid.md [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+### Kernel Switch
+We should `reorder` the different Layout from other device or to other device. `GetExpectedKernelType` and `trans` functions can help us to implement it.
+
+`GetExpectedKernelType` should get the context, and this operator can return the best `KernelType`. 
+`trans` would be like this:
+
+```c++
+void trans(inputs, ctx) override {
+  if (NoNeedTrasn()) {
```
Comment (QiJune) [inline doc/design/mkl/mkldnn_fluid.md]: NoNeedTrans()

File: doc/design/mkl/mkldnn_fluid.md (hunk 1 above)
Comment (tensor-tang) [reply doc/design/mkl/mkldnn_fluid.md]: Got it, thx.

File: doc/design/mkl/mkldnn_fluid.md (hunk 4 above)
Comment (tensor-tang) [reply doc/design/mkl/mkldnn_fluid.md]: My mistake, That is typo. I will corrent it.

File: doc/design/mkl/mkldnn_fluid.md (hunk 2 above)
Comment (tensor-tang) [reply doc/design/mkl/mkldnn_fluid.md]: Sure, I will modify it and I can attach this [link](http://01org.github.io/mkl-dnn) . The example shown is good enough,

File: doc/design/mkl/mkldnn_fluid.md (hunk 3 above)
Comment (tensor-tang) [reply doc/design/mkl/mkldnn_fluid.md]: yeah, thx.

File: doc/design/mkl/mkldnn_fluid.md (hunk 5 above)
Comment (tensor-tang) [reply doc/design/mkl/mkldnn_fluid.md]: Thx I will refine.

File: doc/design/mkl/mkldnn_fluid.md (hunk 6 above)
Comment (tensor-tang) [reply doc/design/mkl/mkldnn_fluid.md]: OK thx.

File: doc/design/mkl/mkldnn_fluid.md (hunk 7 above)
Comment (tensor-tang) [reply doc/design/mkl/mkldnn_fluid.md]: I have added the code in `createPrimitiveDesc`:
```c++
prop_kind pk = ctx.Attr<bool>("is_test") ? prop_kind::forward_inference : prop_kind::forward_training;
```

File: doc/design/mkl/mkldnn_fluid.md (hunk 8 above)
Comment (tensor-tang) [reply doc/design/mkl/mkldnn_fluid.md]: Yes

File: doc/design/mkl/mkldnn_fluid.md (hunk 9 above)
Comment (tensor-tang) [reply doc/design/mkl/mkldnn_fluid.md]: Yes, I get your point, I can not reuse the name. Thanks for catching it!

File: doc/design/mkl/mkldnn_fluid.md (hunk 10 above)
Comment (tensor-tang) [reply doc/design/mkl/mkldnn_fluid.md]: Thanks~

**Comment by tensor-tang** [comment]
Done

File: doc/design/mkl/mkldnn_fluid.md [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+1.  [How to write a new operator](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/howto/dev/new_op_en.md). We are trying to add a new kind of kernel into operators, so basically we should follow this doc.
+2.  [Supporting new Device/Library](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/support_new_device.md). Since MKLDNN is a new library to fluid, we should add `MKLDNNDeviceContext` and maybe `mkldnn_helper.h`, just like [cudnn_helper.h](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/platform/cudnn_helper.h).
+3.  [Switch Kernel](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/switch_kernel.md). Another important point is that we should ensure the data synchronization between different kernel types, which is this [topic](https://github.com/PaddlePaddle/Paddle/issues/6549). So basically we should override `GetExpectedKernelType` and `trans` functions to support switching kernels.
+4.  [The Keys of Operator Kernel Type](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/operator_kernel_type.md). Kernel Type is a pivotal conception which can record the `Place`, `Library`, `DataType` and `Layout`.
+
+## Sulution
```
Comment (luotao1) [inline doc/design/mkl/mkldnn_fluid.md]: Solution

File: doc/design/mkl/mkldnn_fluid.md [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+In general, there are four parts we should follow to run a MKL-DNN primitive.
+-  Create a primitive descriptor that describe this operator
+-  Create a primitive itself by primitive descriptor and the engine
+-  Create all memory buffers that primitive needed
+-  Launch a stream to execute the primitive created
+More details can refer to [here](http://01org.github.io/mkl-dnn).
```
Comment (luotao1) [inline doc/design/mkl/mkldnn_fluid.md]: 17-18行要空一行。

File: doc/design/mkl/mkldnn_fluid.md [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+It's better to avoid reinitialization of primitives and memory handles in the first three stages in every iteration. \
+So we plan to create a map to record all the `primitive` and `memory`, which should not take too much memories as discussed [here](https://github.com/PaddlePaddle/Paddle/issues/6822).
+
+It's assumed that following three conditions should be satisfied.
+1. there is a unique key for each operator instance. May be the actual name of `Output Tensor`.
+2. the `Input Tensor` inside `Compute` function is the one after converted.
+3. we can get the phase(eg. `is_test`) inside `Compute` function, otherwise we need to expose this attribue to user.
+
+### Compute
+The algorithm of `Compute` would be described as follow, let's take conv like an example.
```
Comment (luotao1) [inline doc/design/mkl/mkldnn_fluid.md]: `conv`

File: doc/design/mkl/mkldnn_fluid.md [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+-  Create a primitive descriptor that describe this operator
+-  Create a primitive itself by primitive descriptor and the engine
+-  Create all memory buffers that primitive needed
+-  Launch a stream to execute the primitive created
+More details can refer to [here](http://01org.github.io/mkl-dnn).
+
+It's better to avoid reinitialization of primitives and memory handles in the first three stages in every iteration. \
+So we plan to create a map to record all the `primitive` and `memory`, which should not take too much memories as discussed [here](https://github.com/PaddlePaddle/Paddle/issues/6822).
+
+It's assumed that following three conditions should be satisfied.
```
Comment (luotao1) [inline doc/design/mkl/mkldnn_fluid.md]: Assume that
